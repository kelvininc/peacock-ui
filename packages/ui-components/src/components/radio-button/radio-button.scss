@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --button-height: Radio button component's height.
	 * @prop --text-color-default: Radio button component's text color default.
	 * @prop --text-color-active: Radio button component's text color active.
	 * @prop --text-color-disabled: Radio button component's text color disabled.
	 * @prop --background-color-default: Radio button component's background color default.
	 * @prop --background-color-active: Radio button component's background color active.
	 * @prop --background-color-disabled: Radio button component's background color disabled.
	 * @prop --border-color-default: Radio button component's border color default.
	 * @prop --border-color-active: Radio button component's border color active.
	 * @prop --border-color-disabled: Radio button component's border color disabled.
	 */

	--button-height: 36px;
	--text-color-default: #{kv-color('neutral-2')};
	--text-color-active: #{kv-color('neutral-0')};
	--text-color-disabled: #{kv-color('neutral-4')};
	--background-color-default: #{kv-color('neutral-6')};
	--background-color-active: #{kv-color('neutral-6')};
	--background-color-disabled: #{kv-color('neutral-6')};
	--border-color-default: #{kv-color('neutral-6')};
	--border-color-active: #{kv-color('neutral-0')};
	--border-color-disabled: #{kv-color('neutral-6')};
}

.radio-button {
	@include kv-font-label-small-regular();

	text-decoration: none;

	position: relative;
	user-select: none;
	cursor: pointer;

	display: inline-flex;
	align-items: center;

	border-radius: 2px;

	border: 1px solid var(--border-color-default);
	color: var(--text-color-default);
	background: var(--background-color-default);

	height: var(--button-height);
	padding: 0 $spacing-3x;

	&--checked {
		text-shadow: 0 0 0.35px var(--text-color-active), 0 0 0.35px var(--text-color-active);

		border-color: var(--border-color-active);
		color: var(--text-color-active);
		background: var(--background-color-active);
	}

	&--disabled {
		border-color: var(--border-color-disabled);
		color: var(--text-color-disabled);
		background: var(--background-color-disabled);

		cursor: not-allowed;
	}
}
