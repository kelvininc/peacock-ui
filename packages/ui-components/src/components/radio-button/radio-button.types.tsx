import { EventEmitter } from '@stencil/core';

export interface IRadioButton {
	/** (required) Adds a label aside the button */
	label: string;
	/** (optional) The value to be emitted upon click events */
	value?: string;
	/** (optional) Sets the button's styling to be disabled and disables click events */
	disabled?: boolean;
	/** (optional) Sets the button as checked */
	checked?: boolean;
}

export interface IRadioButtonEvents {
	/** Emits when a button is clicked */
	checkedChange: EventEmitter<string>;
}
