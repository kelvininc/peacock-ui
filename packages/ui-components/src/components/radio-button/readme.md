# _<kv-radio-button>_

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-radio-button label="Option 1" value="option-1"></kv-radio-button>

<!-- Disabled -->
<kv-radio-button label="Option 1" value="option-1" disabled></kv-radio-button>

<!-- Checked -->
<kv-radio-button label="Option 1" value="option-1" checked></kv-radio-button>
```


### React

```tsx
import React from 'react';

import { KvRadioButton } from '@kelvininc/react-ui-components';

export const RadioButtonExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvRadioButton label="Option 1" value="option-1" />

		{/*-- Disabled --*/}
		<KvRadioButton label="Option 1" value="option-1" disabled />

		{/*-- Checked --*/}
		<KvRadioButton label="Option 1" value="option-1" checked />
	</>
);
```



## Properties

| Property             | Attribute  | Description                                                                   | Type      | Default      |
| -------------------- | ---------- | ----------------------------------------------------------------------------- | --------- | ------------ |
| `checked`            | `checked`  | (optional) Sets the button as checked                                         | `boolean` | `false`      |
| `disabled`           | `disabled` | (optional) Sets the button's styling to be disabled and disables click events | `boolean` | `false`      |
| `label` _(required)_ | `label`    | (required) Adds a label aside the button                                      | `string`  | `undefined`  |
| `value`              | `value`    | (optional) The value to be emitted upon click events                          | `string`  | `this.label` |


## Events

| Event           | Description                    | Type                  |
| --------------- | ------------------------------ | --------------------- |
| `checkedChange` | Emits when a button is clicked | `CustomEvent<string>` |


## Shadow Parts

| Part             | Description       |
| ---------------- | ----------------- |
| `"radio-button"` | The radio action. |


## CSS Custom Properties

| Name                          | Description                                         |
| ----------------------------- | --------------------------------------------------- |
| `--background-color-active`   | Radio button component's background color active.   |
| `--background-color-default`  | Radio button component's background color default.  |
| `--background-color-disabled` | Radio button component's background color disabled. |
| `--border-color-active`       | Radio button component's border color active.       |
| `--border-color-default`      | Radio button component's border color default.      |
| `--border-color-disabled`     | Radio button component's border color disabled.     |
| `--button-height`             | Radio button component's height.                    |
| `--text-color-active`         | Radio button component's text color active.         |
| `--text-color-default`        | Radio button component's text color default.        |
| `--text-color-disabled`       | Radio button component's text color disabled.       |


## Dependencies

### Used by

 - [kv-radio-button-group](../radio-button-group)

### Graph
```mermaid
graph TD;
  kv-radio-button-group --> kv-radio-button
  style kv-radio-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


