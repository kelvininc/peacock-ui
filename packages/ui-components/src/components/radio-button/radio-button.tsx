import { Component, Host, h, Prop, EventEmitter, Event } from '@stencil/core';
import { throttle } from 'lodash-es';
import { DEFAULT_THROTTLE_WAIT } from '../../config';
import { IAnchor, EAnchorTarget } from '../../types';
import { IRadioButton, IRadioButtonEvents } from './radio-button.types';

/**
 * @part radio-button - The radio action.
 */
@Component({
	tag: 'kv-radio-button',
	styleUrl: 'radio-button.scss',
	shadow: true
})
export class KvRadioButton implements IRadioButton, IRadioButtonEvents, IAnchor {
	/** @inheritdoc */
	@Prop({ reflect: true }) label!: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) value: string = this.label;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) checked?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) href?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) target?: EAnchorTarget;
	/** @inheritdoc */
	@Prop({ reflect: true }) download?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) preventDefault? = false;

	/** @inheritdoc */
	@Event() checkedChange: EventEmitter<string>;

	private clickThrottler: (e: MouseEvent) => void;
	private onCheck = () => {
		if (!this.disabled) {
			this.checkedChange.emit(this.value);
		}
	};

	connectedCallback() {
		this.clickThrottler = throttle(() => this.onCheck(), DEFAULT_THROTTLE_WAIT);
	}

	onClick(event: MouseEvent) {
		if (this.preventDefault) {
			event.preventDefault();
		}

		this.clickThrottler(event);
	}

	render() {
		return (
			<Host>
				<a
					class={{
						'radio-button': true,
						'radio-button--checked': this.checked,
						'radio-button--disabled': this.disabled
					}}
					part="radio-button"
					onClick={this.onClick.bind(this)}
					download={this.download}
					href={this.href}
					target={this.target}
				>
					{this.label}
				</a>
			</Host>
		);
	}
}
