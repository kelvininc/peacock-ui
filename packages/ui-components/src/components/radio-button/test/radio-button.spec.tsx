import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KvRadioButton } from '../radio-button';

describe('Radio Button (unit tests)', () => {
	let page: SpecPage;
	let component: KvRadioButton;

	describe('when the component loads with default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRadioButton],
				html: '<kv-radio-button label="Option 1"></kv-radio-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with disabled prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRadioButton],
				html: '<kv-radio-button label="Option 1" disabled></kv-radio-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with checked prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRadioButton],
				html: '<kv-radio-button label="Option 1" checked></kv-radio-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with value prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRadioButton],
				html: '<kv-radio-button label="Option 1" value="opt1"></kv-radio-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with preventDefault prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRadioButton],
				html: '<kv-radio-button label="Option 1" prevent-default="true"></kv-radio-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		describe('and onClick is called', () => {
			let onClickSpyEvent: MouseEvent;

			beforeEach(() => {
				onClickSpyEvent = new MouseEvent('click');
				component.onClick(onClickSpyEvent);
			});

			it('should prevent event', () => {
				expect(onClickSpyEvent.defaultPrevented).toBeTruthy();
			});
		});
	});
});
