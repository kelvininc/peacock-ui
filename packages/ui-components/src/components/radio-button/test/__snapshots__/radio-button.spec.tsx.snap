// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Button (unit tests) when the component loads with checked prop should match the snapshot 1`] = `
<kv-radio-button checked="" label="Option 1" value="Option 1">
  <mock:shadow-root>
    <a class="radio-button radio-button--checked" part="radio-button">
      Option 1
    </a>
  </mock:shadow-root>
</kv-radio-button>
`;

exports[`Radio Button (unit tests) when the component loads with default props should match the snapshot 1`] = `
<kv-radio-button label="Option 1" value="Option 1">
  <mock:shadow-root>
    <a class="radio-button" part="radio-button">
      Option 1
    </a>
  </mock:shadow-root>
</kv-radio-button>
`;

exports[`Radio Button (unit tests) when the component loads with disabled prop should match the snapshot 1`] = `
<kv-radio-button disabled="" label="Option 1" value="Option 1">
  <mock:shadow-root>
    <a class="radio-button radio-button--disabled" part="radio-button">
      Option 1
    </a>
  </mock:shadow-root>
</kv-radio-button>
`;

exports[`Radio Button (unit tests) when the component loads with preventDefault prop should match the snapshot 1`] = `
<kv-radio-button label="Option 1" prevent-default="" value="Option 1">
  <mock:shadow-root>
    <a class="radio-button" part="radio-button">
      Option 1
    </a>
  </mock:shadow-root>
</kv-radio-button>
`;

exports[`Radio Button (unit tests) when the component loads with value prop should match the snapshot 1`] = `
<kv-radio-button label="Option 1" value="opt1">
  <mock:shadow-root>
    <a class="radio-button" part="radio-button">
      Option 1
    </a>
  </mock:shadow-root>
</kv-radio-button>
`;
